From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uazo <uazo@users.noreply.github.com>
Date: Tue, 12 Apr 2022 15:58:01 +0000
Subject: Add support for writing URIs

Allows native-side URI file writing

Original License: GPL-2.0-or-later - https://spdx.org/licenses/GPL-2.0-or-later.html
License: GPL-3.0-only - https://spdx.org/licenses/GPL-3.0-only.html
---
 base/android/content_uri_utils.cc             | 10 ++++++
 base/android/content_uri_utils.h              |  4 +++
 .../org/chromium/base/ContentUriUtils.java    | 32 +++++++++++++++++++
 3 files changed, 46 insertions(+)

diff --git a/base/android/content_uri_utils.cc b/base/android/content_uri_utils.cc
index 82c59768fec02e8f6313e80229954789360d43c2..328148f4df63cfc87966959918b7ddba07aa2a06 100644
--- a/base/android/content_uri_utils.cc
+++ b/base/android/content_uri_utils.cc
@@ -157,6 +157,16 @@ bool IsContentUriFromThisApp(const FilePath& content_uri) {
   return Java_ContentUriUtils_isUriFromThisApp(env, content_uri.value());
 }
 
+File OpenContentUriForWrite(const FilePath& content_uri) {
+  JNIEnv* env = base::android::AttachCurrentThread();
+  ScopedJavaLocalRef<jstring> j_uri =
+      ConvertUTF8ToJavaString(env, content_uri.value());
+  jint fd = Java_ContentUriUtils_openContentUriForWrite(env, j_uri);
+  if (fd < 0)
+    return File();
+  return File(fd);
+}
+
 std::string GetContentUriMimeType(const FilePath& content_uri) {
   JNIEnv* env = android::AttachCurrentThread();
   return Java_ContentUriUtils_getMimeType(env, content_uri.value());
diff --git a/base/android/content_uri_utils.h b/base/android/content_uri_utils.h
index 2a602f32aca6c857b65ed31c52b66e0f782db4cf..fa6bb002ff4c1e448105ad5cd8aa39d8de6b5497 100644
--- a/base/android/content_uri_utils.h
+++ b/base/android/content_uri_utils.h
@@ -46,6 +46,10 @@ int ContentUriGetFd(
 void ContentUriClose(
     const base::android::JavaRef<jobject>& java_parcel_file_descriptor);
 
+// Opens a content URI for write and returns the file descriptor to the caller.
+// Returns -1 if the URI is invalid.
+BASE_EXPORT File OpenContentUriForWrite(const FilePath& content_uri);
+
 // Returns true if file exists and results are populated, else returns false.
 bool ContentUriGetFileInfo(const FilePath& content_uri,
                            FileEnumerator::FileInfo* results);
diff --git a/base/android/java/src/org/chromium/base/ContentUriUtils.java b/base/android/java/src/org/chromium/base/ContentUriUtils.java
index 48943083c9460635b4dcc7a1de1696f68a9c0997..ef3b15f9bc0efe9d2e0b661decf5caa2ce3203f0 100644
--- a/base/android/java/src/org/chromium/base/ContentUriUtils.java
+++ b/base/android/java/src/org/chromium/base/ContentUriUtils.java
@@ -6,6 +6,7 @@ package org.chromium.base;
 
 import static org.chromium.build.NullUtil.assumeNonNull;
 
+import android.content.ContentProviderClient;
 import android.content.ContentResolver;
 import android.content.Context;
 import android.content.pm.PackageManager;
@@ -18,6 +19,7 @@ import android.os.ParcelFileDescriptor;
 import android.provider.DocumentsContract;
 import android.provider.MediaStore;
 import android.provider.OpenableColumns;
+import android.system.Os;
 import android.text.TextUtils;
 import android.webkit.MimeTypeMap;
 
@@ -79,6 +81,36 @@ public abstract class ContentUriUtils {
         StreamUtil.closeQuietly(parcelFileDescriptor);
     }
 
+    @CalledByNative
+    public static int openContentUriForWrite(String uriString) {
+        try {
+            Uri uri = Uri.parse(uriString);
+            ContentResolver resolver = ContextUtils.getApplicationContext().getContentResolver();
+            ContentProviderClient client =
+                    resolver.acquireContentProviderClient(uri.getAuthority());
+            ParcelFileDescriptor pfd = client.openFile(uri, "wt");
+            int fd = pfd.detachFd();
+            client.close();
+            return fd;
+        } catch (Exception e) {
+            Log.e(TAG, "Cannot open intermediate URI", e);
+        }
+        return -1;
+    }
+
+    public static String getFilePathFromContentUri(Uri uri) {
+        String path = null;
+        try {
+            ContentResolver resolver = ContextUtils.getApplicationContext().getContentResolver();
+            ParcelFileDescriptor pfd = resolver.openFileDescriptor(uri, "r");
+            path = Os.readlink("/proc/self/fd/" + pfd.getFd());
+            pfd.close();
+        } catch (Exception e) {
+            Log.e(TAG, "Cannot get file path from content URI", e);
+        }
+        return path;
+    }
+
     /**
      * Check whether a content URI exists.
      *
